### ISSUE TITLE
Respect Keras layer names for output operations in Concrete Function

### ISSUE THREAD

**Issue by DLumi** [keras-team/keras#18012]
**System information**.

TensorFlow version (you are using): 2.12.0
Are you willing to contribute it (Yes/No) : No

**Describe the feature and the current behavior/state**.

When converting a Keras model to concrete function, you can preserve the input name by creating a named TensorSpec, but the outputs are always created for you by just slapping tf.identity on top of whatever you had there, even if it was a custom named tf.identity operation. Since many converters rely on concrete functions to make their own representation (TFLite, ONNX, CoreML, etc), this behavior messes up the output operation names, often making them inconsistent with each other.
There's currently no workaround for that. You can access previous graph nodes by calling a layer named like {model_name}/{output_layer_name} when doing inference on frozen graph itself, but it won't help you in any way to convert the model.

So I'd be happy to see one of those things as a solution to that:

Add an option to explicitly specify the TensorSpec for outputs, just the way we do it for inputs. This would be the most obvious and convenient way of doing it from a user standpoint
Don't add new identity operations on top of existing ones. More of a kludge, but would get the job done
Add an option to rename operations in concrete function post factum.
Add an option to cut off the operations in concrete function past a certain node.
Add an option to convert a graph into a concrete function. Since you can directly modify graphs, this could work as well

**Will this change the current api? How?**

Depends on the implementation

**Who will benefit from this feature?**

Anyone who cares about the output names in concrete function and basically every other external format (since the model typically gets frozen prior to conversion)

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): No
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing): N/A

**Comment by DLumi** [comment]
This is a duplicate directed to Keras team as suggested by @sushreebarsa in TensorFlow repo:
https://github.com/tensorflow/tensorflow/issues/60289#issuecomment-1511611448

**Comment by tilakrayal** [comment]
@DLumi,
Could you please specify the Use Cases for this feature, it helps us to debug the issue in an effective way. Thank you!

**Comment by DLumi** [comment]
> @DLumi, Could you please specify the Use Cases for this feature, it helps us to debug the issue in an effective way. Thank you!

As I mentioned in my original message, this would introduce manual controll over output names in Concrete Functions, thus it'd help to preserve output names in other formats like TFLite, ONNX, CoreML, etc, that internally rely on freeszing the model prior to conversion. 
Basically, this would help with input/output names consistency, while keeping the names unique and distinguishable (Identity, Identity_1, Identity_2, etc - is not what I call 'distinguishable').

**Comment by DLumi** [comment]
> > @DLumi, Could you please specify the Use Cases for this feature, it helps us to debug the issue in an effective way. Thank you!
> 
> As I mentioned in my original message, this would introduce manual controll over output names in Concrete Functions, thus it'd help to preserve output names in other formats like TFLite, ONNX, CoreML, etc, that internally rely on freezing the model prior to conversion. Basically, this would help with input/output names consistency, while keeping the names unique and distinguishable (Identity, Identity_1, Identity_2, etc - is not what I'd call 'distinguishable').

**Comment by mihirparadkar** [comment]
Hi, I think this problem is better suited for the export archive functionality available in https://keras.io/api/models/model_saving_apis/export/ for fine-grained control of the endpoint names.

**Comment by sachinprasadhs** [comment]
@DLumi , Could you please refer the above comment for ideal usage.
Also, below is the workaround for using identity layer.

```
scores = tf.keras.layer.Identity(scores, name="scores)
boxes = tf.keras.layer.Identity(boxes, name="boxes)

model = tf.keras.Model({"input": input_tensor}, {"boxes": boxes, "scores": scores})
```

**Comment by DLumi** [comment]
> @DLumi , Could you please refer the above comment for ideal usage. Also, below is the workaround for using `Identity` layer.
> 
> ```
> scores = tf.keras.layer.Identity(scores, name="scores)
> boxes = tf.keras.layer.Identity(boxes, name="boxes)
> 
> model = tf.keras.Model({"input": input_tensor}, {"boxes": boxes, "scores": scores})
> ```

I don't think I fully get the idea behind the `ExportArchive` or its usecases for that matter. As far as I understand, it is a fancier way to write SavedModel artifacts there. But I don't wanna write anything, I need an in-memory object which would be able to produce a concrete function with adequate (i.e. corresponding to Keras) names. And as of right now I don't think you can specify output names there anyway. 

As for workaround you suggested, well, it's not working, and I'm pretty sure I mentioned this one in my original post. When you request a ConcreteFunction, the algorithm just slaps a fresh copy of `tf.identity` on top of whatever layer/operation you had there. Here's the code te reproduce, so you can check it yourself if you wish. 

```python
import tensorflow as tf

# Build simple model
inputs = tf.keras.Input((224, 224, 3), name='custom_input_layer')
x = tf.keras.layers.Flatten()(inputs)
x = tf.keras.layers.Dense(512, activation='relu')(x)
x = tf.keras.layers.Dense(256, activation='relu')(x)
x = tf.keras.layers.Dense(128, activation='relu')(x)
x = tf.keras.layers.Dense(1, activation='sigmoid', name='custom_output_layer')(x)
x = tf.keras.layers.Identity(name='custom_output_layer_id')(x)

model = tf.keras.Model(inputs=inputs, outputs=x, name='my_custom_model_name')
model.summary()

input_tensors = [tf.TensorSpec(shape=inp.shape, dtype=tf.float32, name=inp.name) for inp in model.inputs]
concrete_function = tf.function(lambda x: model(x)).get_concrete_function(x=input_tensors)

print(concrete_function.outputs)  # note that the node is called Identity, not `custom_output_layer_id`
```

**Comment by ianstenbit** [comment]
I tend to agree with your [comment](https://github.com/tensorflow/tensorflow/issues/60289#issuecomment-1512357816) that this is a TensorFlow issue, not a Keras issue.

From your last comment's example code, the resulting model's `model.output` has an appropriate name, and I also tried using a direct `tf.identity(x, name="...")` instead of `tf.keras.layers.Identity`, with the same result.

This is fundamentally an issue with `get_concrete_function`, which is a TF construct, not Keras. Unfortunately, I don't think there's a viable workaround in Keras itself because Keras is already producing the expected named outputs which just get lost in translation.

@DLumi sorry that we've bounced you between repos a bit, but I agree with you that the issue should stay open in TF and not here.
